In [317]:
import pandas as pd
import ast
import pytz
import os
import json
import yaml
from dateutil import tz
from datetime import datetime
from datetime import timedelta
from IPython.display import display

pd.set_option('display.max_columns', None)

class BasePipeline:
    def __init__(self, config, base_file_path):
        self.config = config
        self.base_file_path = base_file_path
        self.map = {}
        self._data = pd.DataFrame({})
    
    def construct_file_path(self):
        #TODO: add type to prefix mapping
        file_prefix = 'per_outage' if self.config['type'] == 'o' else 'per_county'
        file_path = f"{self.base_file_path}/{self.config['state']}/layout_{self.config['layout']}/{file_prefix}_{self.config['name']}.csv"
        return file_path.replace('//', '/')

    def load_data(self):
        # TODO: use us zipcode database
        try:
            file_path = self.construct_file_path()
            print(file_path)
            self._data = pd.read_csv(file_path)
            with open(f"{self.config['state']}_mapping.json", 'r') as json_file:
                self.map = json.load(json_file)
        except Exception as e:
            print(f"An error occurred during file loading: {e}")
            
    def transform(self):
        # Base transformation method
        raise NotImplementedError

    def standardize(self):
        # Base transformation method
        # where most of the functionality is
        self.load_data()
        self.transform()
        grouped = self._data.groupby('outage_id').apply(self._compute_metrics).reset_index().round(2)
        self._data = pd.merge(grouped, self._data, on=['outage_id', 'timestamp'], how='inner')
        
        self._data['utility_provider'] = self.config['name']
        self._data['state'] = self.config['state']
        self._data['county'] = self._data['zipcode'].map(self.map)
        
        self._data = self._data[[
            'utility_provider', 'state', 'county', 'zipcode',
            'outage_id', 'start_time', 'end_time', 'lat', 'lng', 
            'duration', 'duration_max', 'duration_mean', 'customer_affected_mean', 'total_customer_outage_time', 'total_customer_outage_time_max', 'total_customer_outage_time_mean'
        ]]
        
        return self._data
    
    def output_data(self, standard_data):
        # TODO: Output unified data
        pass
    
    def get_dataframe(self):
        return self._data
    
    def _compute_metrics(self, group):
        duration = (group['end_time'] - group['start_time']).dt.total_seconds() / 60
        duration_max = duration + 15
        duration_mean = (duration + duration_max) / 2
        customer_affected_mean = group['customer_affected'].mean()
        
        total_customer_outage_time = 15 * (group['customer_affected'].sum() - group['customer_affected'].iloc[0]) + (group['timestamp'].iloc[0] - group['start_time'].iloc[0]).total_seconds() / 60 * group['customer_affected'].iloc[0]
        total_customer_outage_time_max = total_customer_outage_time + 15 * group['customer_affected'].iloc[-1]
        total_customer_outage_time_mean = (total_customer_outage_time + total_customer_outage_time_max) / 2

        return pd.Series({
            'timestamp': group['end_time'].iloc[-1],
            'duration': duration.iloc[-1],
            'duration_max': duration_max.iloc[-1],
            'duration_mean': duration_mean.iloc[-1],
            'customer_affected_mean': customer_affected_mean,
            'total_customer_outage_time': total_customer_outage_time,
            'total_customer_outage_time_max': total_customer_outage_time_max,
            'total_customer_outage_time_mean': total_customer_outage_time_mean
        })
        
    def _check_other_vars(self):
        # TODO: Check other useful variables
        pass
    
    
class GA11TX12(BasePipeline):    
    def __init__(self, config, base_file_path):
        super().__init__(config, base_file_path)
        with open('us_mapping.json', 'r') as json_file:
            self._usmap = json.load(json_file)
    
    def transform(self): # if edited, must recreate pipeline to reset transformed flag
        #### HELPER METHOD
        def _reformat_start_date(row): # taking the row and reformating its 'start_date' based on timestamp
            # Split the date string into components
            # start_date format: 03/15 05:28 pm
            month_day, time, ampm = row['start_date'].split(' ') # taking the row's start_date and parsing it
            # Split the month and day and determine year
            s_month, s_day = month_day.split('/')
            year = None
            # Determining year using timestamp as start_date does not include year
            if pd.notna(row['timestamp']): # if the timestamp value for the given row is not null (hopefully it works)
            # timestamp format: 01-18-2024 15:25:06
                timestamp_components = row['timestamp'].split(' ')
                ts_date_comp = timestamp_components[0].split('-')
                t_month, t_day = ts_date_comp[0], ts_date_comp[1]
                t_year = pd.to_numeric(ts_date_comp[2])
                if t_month == '01' and s_month == '12':
                    # if the timestamp is in january but the start time is reported to be in december, make the update time have the previous year (year of Jan - 1)
                    year = str(int(t_year) - 1)
                else: # use timestamp year
                    year = t_year 
            else: # if no timestamp
            # for Walton, Tri-State, Oconee, and Mitchell, the na timestamps are in march 2023
                year = '2023'

            # Extract the hour and minute from the time
            hour, minute = time.split(':')

            # Convert hour to 24-hour format
            if 'am' in ampm.lower() and hour == '12': # if 12 am, then set to 00
                hour = '00' 
            if 'pm' in ampm.lower():
                hour = str(int(hour) + 12) if int(hour) < 12 else hour

            # Add leading zeros if necessary
            hour = hour.zfill(2)
            minute = minute.zfill(2)

            # Combine the components into the desired format
            reformatted_date = f'{s_month}-{s_day}-{year} {hour}:{minute}:00'

            return reformatted_date


        # Helper method
        def _reformat_update(row):
            # Format: Mar 15, 5 09, pm
            ## Splitting updateTime into components month, day, hour, min, am/pm
            month_day, time, ampm = row['updateTime'].split(',') # split into date; time; and pm/am 
            # Splitting into month, day
            u_month, u_day = month_day.split(' ')
            month_dict = { 'Jan': '01', 'Feb': '02', 'Mar': '03', 'Apr': '04', 'May': '05', 'Jun': '06', 
                        'Jul': '07', 'Aug': '08', 'Sep': '09', 'Oct': '10', 'Nov': '11', 'Dec': '12' }
            u_month = month_dict[u_month] # turning the 3 char month name to its two digit form
            
            # Determining year from timestamp as updateTime does not have year
            year = None
            if pd.notna(row['timestamp']): # if the timestamp value for the given row is not null (hopefully it works)
                # timestamp format: 01-18-2024 15:25:06
                timestamp_components = row['timestamp'].split(' ')
                ts_date_comp = timestamp_components[0].split('-')
                t_month, t_day = ts_date_comp[0], ts_date_comp[1]
                t_year = pd.to_numeric(ts_date_comp[2])

                if t_month == '01' and u_month == '12':
                    # if the timestamp is in january but the start time is reported to be in december, make the update time have the previous year (year of Jan - 1)
                    year = str(int(t_year) - 1)
                else: # use timestamp year
                    year = t_year 
            else: # if no timestamp
            # for Walton, Tri-State, Oconee, and Mitchell, the na timestamps are in march 2023
                year = '2023'

            # Extract the hour and minute from the update time
            hour, minute = time.split() # splits based on whtie space which is space in this case and removed leading space
            # Convert hour to 24-hour format
            if 'am' in ampm.lower() and hour == '12': # if 12 am, then set to 00
                hour = '00' 
            if 'pm' in ampm.lower():
                hour = str(int(hour) + 12) if int(hour) < 12 else hour # if pm and less than 12, add 12 hours  --> othertest it is 12 pm

            # Add leading zeros if necessary
            hour = hour.zfill(2)
            minute = minute.zfill(2)

            # Combine the components into the desired format
            reformatted_date = f'{u_month}-{u_day}-{year} {hour}:{minute}:00'

            return reformatted_date

        ## HELPER METHOD ####
        

        try:
            self._data['start_date'] = self._data.apply(_reformat_start_date, axis=1) # reformattin the string to be datetime string form
            self._data['start_date'] = pd.to_datetime(self._data['start_date']) # change to timestamp data format
            self._data['updateTime'] = self._data.apply(_reformat_update, axis=1)
            self._data['updateTime'] = pd.to_datetime(self._data['updateTime'])
            self._data['duration'] = pd.to_timedelta(self._data['duration'])
            self._data['timestamp'] = pd.to_datetime(self._data['timestamp']) # THIS HAS TO BE BEFORE applying reformat functions as splitting only works on timestamp in string form, not datetime

            # Renaming column names to match with superclass standardize
            self._data = self._data.rename(columns={
                'incident_id':'outage_id',
                'zip_code':'zipcode'
            })
            
            # Eliminating outage_id's with multiple start dates since there are not that many of them
            df = self._data
            grouped = df.groupby('outage_id')['start_date'].nunique()
            multi_start_outages = grouped[grouped > 1].index.tolist()
            self._data = df[~df['outage_id'].isin(multi_start_outages)]


        except Exception as e:
            print(f"An error occurred during transformation: {e}")

    def standardize(self):
        # Specific transformation for GA11TX12
        # print(self.config)
        self.load_data()
        self.transform()
        grouped = self._data.groupby('outage_id').apply(self._compute_metrics).reset_index().round(2)
        self._data = grouped

    def _compute_metrics(self, group): # overwriting super class because the most accurate duration seems to be calculated from update times
        # group = groupby groupy of a unique outage_id
                
        duration_diff = group['duration'].max()
        duration_max = duration_diff + timedelta(minutes=15) # because 15 minute update intervals
        duration_mean = (duration_diff + duration_max) / 2
        start_time = group['start_date'].iloc[0]
        end_time = start_time + duration_diff
        customer_affected_mean = group['consumers_affected'].mean()
        total_customer_outage_time = customer_affected_mean * duration_diff
        zipcode = group['zipcode'].iloc[-1]
        zipcode_values = None
        
        null_zipcode = [None, None, None] 
        try:
            zipcode_values = self.map[zipcode] # the tuple of values from zipcode map (county name, fip, state) 
        except KeyError:
            try:
                zipcode_values = self._usmap[zipcode]
            except KeyError:
                # print(f"Nonexistent zipcode in {self.config['name']}: {zipcode}")        
                zipcode_values = null_zipcode


        return pd.Series({
            'start_time': start_time,
            'end_time': end_time,
            'lat': group['lat'].iloc[-1],
            'long': group['lon'].iloc[-1],
            'zipcode': zipcode,
            'county_name': zipcode_values[0], 
            'county_fips': zipcode_values[1],
            'utility_provider': self.config['name'],
            'state': zipcode_values[2],
            'duration_diff': duration_diff,
            'duration_max': duration_max,
            'duration_mean': duration_mean,
            'customer_affected_mean': customer_affected_mean,
            'total_customer_outage_time': total_customer_outage_time
        })


One Time Code to Generate the mappings from US_zips


In [318]:
# us_zip_path = '/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/US Zipcode Data/uszips.csv'
# with open(us_zip_path, 'r') as us_zip:
#     us_zip_df = pd.read_csv(us_zip)
#     us_zip_df.to_csv('filename.csv', index=False)

      # # Creating a US zip file  
#     us_zip_dict = dict(zip(us_zip_df['zip'], zip(us_zip_df['county_name'], us_zip_df['county_fips'], us_zip_df['state_name'])))
#     with open("us_mapping.json", 'w') as json_file:
#         json.dump(us_zip_dict, json_file)


# # Creating GA mapping
# states = ['Georgia', 'Texas'] # states to generate mappings for (update as needed)
# state_abbr_map = {
#     'Georgia': 'ga',
#     'Texas': 'tx'
# }
# for state in states:
#     state_df = us_zip_df[us_zip_df['state_name'] == state]
#     state_zip_dict = dict(zip(state_df['zip'], zip(state_df['county_name'], state_df['county_fips'], state_df['state_name'])))
#     with open(f"{state_abbr_map[state]}_mapping.json", 'w') as json_file:
#         json.dump(state_zip_dict, json_file)

Establishing config files

In [319]:
local_config_path = '/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/app/pipeline/config.yaml'
with open(local_config_path, 'r') as file:
    config = yaml.safe_load(file)
    base_file_path = config['globals']['local_base_file_path']

# Test Code to Feel out and Validate Dataframe

Loading configs manually and separately

In [320]:
config_list = [
    (walton_conf := {'name': 'Walton EMC', 'state': 'ga', 'layout': 11, 'type': 'o'}),
    (tris_conf := {'name': 'Tri-State EMC', 'state': 'ga', 'layout': 11, 'type': 'o'}),
    (oconee_conf := {'name': 'Oconee EMC', 'state': 'ga', 'layout': 11, 'type': 'o'}),
    (mitchell_conf := {'name': 'Mitchell EMC', 'state': 'ga', 'layout': 11, 'type': 'o'}),
    (wise_conf := {'name': 'Wise Electric Coop, Inc.', 'state': 'tx', 'layout': 12, 'type': 'o'}),
    (houston_conf := {'name': 'Houston County Electric Coop, Inc.', 'state': 'tx', 'layout': 12, 'type': 'o'}),
    (cherokee_conf := {'name': 'Cherokee County Electric Coop Association', 'state': 'tx', 'layout': 12, 'type': 'o'})
]

pipelines = [
    (walton_pipe := GA11TX12(walton_conf, base_file_path)),
    (tris_pipe := GA11TX12(tris_conf, base_file_path)),
    (oconee_pipe := GA11TX12(oconee_conf, base_file_path)),
    (mitchell_pipe := GA11TX12(mitchell_conf, base_file_path)),
    (wise_pipe := GA11TX12(wise_conf, base_file_path)),
    (houston_pipe := GA11TX12(houston_conf, base_file_path)),
    (cherokee_pipe := GA11TX12(cherokee_conf, base_file_path))
]

for pipeline in pipelines:
    pipeline.load_data()
    
# for pipeline in pipelines:
#     pipeline.transform()

dataframes = [
    (walton_df := walton_pipe._data),
    (tris_df := tris_pipe._data),
    (oconee_df := oconee_pipe._data),
    (mitchell_df := mitchell_pipe._data),
    (wise_df := wise_pipe._data),
    (houston_df := houston_pipe._data),
    (cherokee_df := cherokee_pipe._data)
]

display(walton_df)

/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Walton EMC.csv


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:32: DtypeWarning: Columns (20,21) have mixed types. Specify dtype option on import or set low_memory=False.
  self._data = pd.read_csv(file_path)


/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Tri-State EMC.csv
/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Oconee EMC.csv
/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Mitchell EMC.csv


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:32: DtypeWarning: Columns (5,6,21) have mixed types. Specify dtype option on import or set low_memory=False.
  self._data = pd.read_csv(file_path)


/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/tx/layout_12/per_outage_Wise Electric Coop, Inc..csv
/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/tx/layout_12/per_outage_Houston County Electric Coop, Inc..csv
/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/tx/layout_12/per_outage_Cherokee County Electric Coop Association.csv


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:32: DtypeWarning: Columns (5,6) have mixed types. Specify dtype option on import or set low_memory=False.
  self._data = pd.read_csv(file_path)


substation  feeder incident_id  estimated_restore_time  formatted_ert  \
0            14.0     3.0     C950581                     NaN            NaN   
1             1.0     3.0     C950582                     NaN            NaN   
2             1.0     3.0     C950582                     NaN            NaN   
3             1.0     3.0     C950582                     NaN            NaN   
4            14.0     1.0     C950583                     NaN            NaN   
...           ...     ...         ...                     ...            ...   
71389        11.0     5.0    D1009642                     NaN            NaN   
71390         6.0     1.0    C1009637                     NaN            NaN   
71391        34.0     3.0    C1009640                     NaN            NaN   
71392         4.0     2.0    C1009654                     NaN            NaN   
71393         1.0     2.0    C1009665                     NaN            NaN   

           start_date      duration  consumers_affected        lon        lat  \
0      03/15 04:28 pm  00 hr 11 min                   1 -83.669285  33.911106   
1      03/15 04:48 pm  00 hr 05 min                   1 -83.703931  33.760364   
2      03/15 04:48 pm  00 hr 21 min                   1 -83.703931  33.760364   
3      03/15 04:48 pm  00 hr 35 min                   1 -83.703931  33.760364   
4      03/15 05:14 pm  00 hr 10 min                   1 -83.768580  33.900673   
...               ...           ...                 ...        ...        ...   
71389  01/18 08:15 am  01 hr 54 min                   3 -83.440692  33.872791   
71390  01/18 07:37 am  02 hr 32 min                   1 -84.043121  33.810690   
71391  01/18 08:08 am  02 hr 00 min                   1 -83.435911  33.739100   
71392  01/18 09:28 am  00 hr 41 min                   1 -83.512373  33.571093   
71393  01/18 09:47 am  00 hr 22 min                   1 -83.646601  33.785259   

      service_index_name  outages  NumConsumers zip_code  isHighTraffic  \
0               Electric        1        137416    30656          False   
1               Electric        1        137416    30655          False   
2               Electric        1        137416    30655          False   
3               Electric        2        137416    30655          False   
4               Electric        2        137416    30620          False   
...                  ...      ...           ...      ...            ...   
71389           Electric       15        139225    30677          False   
71390           Electric       15        139225    30039          False   
71391           Electric       15        139225    30638          False   
71392           Electric       15        139225    30650          False   
71393           Electric       15        139225    30655          False   

             updateTime         EMC                    alias outage_comment  \
0      Mar 15, 3 39, pm  Walton EMC                      NaN            NaN   
1      Mar 15, 3 54, pm  Walton EMC                      NaN            NaN   
2      Mar 15, 4 09, pm  Walton EMC                      NaN            NaN   
3      Mar 15, 4 24, pm  Walton EMC                      NaN            NaN   
4      Mar 15, 4 24, pm  Walton EMC                      NaN            NaN   
...                 ...         ...                      ...            ...   
71389  Jan 18, 9 10, am  Walton EMC  EXPERIMENT STAT/ 15 KVA            NaN   
71390  Jan 18, 9 10, am  Walton EMC                      NaN            NaN   
71391  Jan 18, 9 10, am  Walton EMC                      NaN            NaN   
71392  Jan 18, 9 10, am  Walton EMC                      NaN            NaN   
71393  Jan 18, 9 10, am  Walton EMC                      NaN            NaN   

                 timestamp opt_code  \
0                      NaN      NaN   
1                      NaN      NaN   
2                      NaN      NaN   
3                      NaN      NaN   
4                      NaN  

Seeing what columns have mostly null values

In [321]:
null_rate = 0
print("Average null value rate per column across all providers")
for dataframe in dataframes:
    null_rate = null_rate + dataframe.isnull().sum() / len(dataframe)

print(null_rate / len(dataframes))


Average null value rate per column across all providers
EMC                       0.000000
NumConsumers              0.000000
alias                     0.710222
consumers_affected        0.000000
duration                  0.000000
estimated_restore_time    0.945823
feeder                    0.017576
formatted_ert             0.945823
incident_id               0.000000
isHighTraffic             0.000000
lat                       0.019902
lon                       0.019902
opt_code                       NaN
outage_comment            0.911656
outages                   0.000000
poly                           NaN
service_index_name        0.000000
start_date                0.000000
substation                0.016832
timestamp                 0.009554
updateTime                0.000000
zip_code                  0.000000
dtype: float64


Seeing the rate of null values, we see that across all the providers:
- outage_comment
- estimated_restore_time
- formatted ert
- alias

have significantly high % of nulls. Thus we will not use them.
Other unused variables:

- feeder
- isHighTraffic
- poly
- opt_code
- substation

### Sample Views of Transformation, Validation, and Compute Metrics

In [322]:
# Control which df to test on
test_conf = mitchell_conf 
test_p = GA11TX12(test_conf, base_file_path)
test_p.load_data()


/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Mitchell EMC.csv


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:32: DtypeWarning: Columns (5,6,21) have mixed types. Specify dtype option on import or set low_memory=False.
  self._data = pd.read_csv(file_path)


#### Issues with Raw Data and Validating Transform:
- The updateTime and start_date do not have the year included, so we need to utilize the timestamp to calculate the years in order to convert to a datetime format to calculate metrics
- However, some rows have null timestamps.

In [323]:
# Display the rows of each dataframe where timestamp is null
for dataframe in dataframes:
    display(dataframe[dataframe['timestamp'].isna()][['incident_id', 'start_date', 'updateTime', 'timestamp']])

incident_id      start_date        updateTime timestamp
0        C950581  03/15 04:28 pm  Mar 15, 3 39, pm       NaN
1        C950582  03/15 04:48 pm  Mar 15, 3 54, pm       NaN
2        C950582  03/15 04:48 pm  Mar 15, 4 09, pm       NaN
3        C950582  03/15 04:48 pm  Mar 15, 4 24, pm       NaN
4        C950583  03/15 05:14 pm  Mar 15, 4 24, pm       NaN
...          ...             ...               ...       ...
1735     C951589  03/29 03:09 pm  Mar 29, 4 09, pm       NaN
1736     C951590  03/29 04:46 pm  Mar 29, 4 09, pm       NaN
1737     C951589  03/29 03:09 pm  Mar 29, 4 24, pm       NaN
1738     C951589  03/29 03:09 pm  Mar 29, 4 39, pm       NaN
1739     C951591  03/29 05:24 pm  Mar 29, 4 39, pm       NaN

[1740 rows x 4 columns]

incident_id      start_date         updateTime timestamp
0       D760313  03/16 09:34 pm   Mar 16, 8 39, pm       NaN
1       D760313  03/16 09:34 pm   Mar 16, 8 54, pm       NaN
2       D760313  03/16 09:34 pm   Mar 16, 9 09, pm       NaN
3       D760353  03/19 09:16 am   Mar 19, 8 24, am       NaN
4       D760353  03/19 09:16 am   Mar 19, 8 39, am       NaN
..          ...             ...                ...       ...
202     D760657  03/28 11:19 am  Mar 28, 12 09, pm       NaN
203     D760660  03/28 01:59 pm   Mar 28, 1 09, pm       NaN
204     D760660  03/28 01:59 pm   Mar 28, 1 24, pm       NaN
205     D760660  03/28 01:59 pm   Mar 28, 1 39, pm       NaN
206     D760660  03/28 01:59 pm   Mar 28, 1 54, pm       NaN

[207 rows x 4 columns]

incident_id      start_date        updateTime timestamp
0        C434216  03/15 04:31 pm  Mar 15, 3 40, pm       NaN
1        C434216  03/15 04:31 pm  Mar 15, 3 55, pm       NaN
2        C434216  03/15 04:31 pm  Mar 15, 4 10, pm       NaN
3        C434216  03/15 04:31 pm  Mar 15, 4 25, pm       NaN
4        C434219  03/16 07:42 pm  Mar 16, 6 55, pm       NaN
...          ...             ...               ...       ...
1140     D434618  03/29 04:52 pm  Mar 29, 4 10, pm       NaN
1141     D434617  03/29 04:35 pm  Mar 29, 4 25, pm       NaN
1142     D434618  03/29 04:52 pm  Mar 29, 4 25, pm       NaN
1143     D434617  03/29 04:35 pm  Mar 29, 4 40, pm       NaN
1144     D434618  03/29 04:52 pm  Mar 29, 4 40, pm       NaN

[1145 rows x 4 columns]

incident_id      start_date        updateTime timestamp
0        G437006  03/15 07:14 am  Mar 15, 3 25, pm       NaN
1        C483513  03/15 02:24 pm  Mar 15, 3 25, pm       NaN
2        G437006  03/15 07:14 am  Mar 15, 3 40, pm       NaN
3        G437006  03/15 07:14 am  Mar 15, 3 55, pm       NaN
4        G437006  03/15 07:14 am  Mar 15, 4 10, pm       NaN
...          ...             ...               ...       ...
2262     G437006  03/29 07:31 am  Mar 29, 3 40, pm       NaN
2263     G437006  03/29 07:31 am  Mar 29, 3 55, pm       NaN
2264     G437006  03/29 07:31 am  Mar 29, 4 10, pm       NaN
2265     G437006  03/29 07:31 am  Mar 29, 4 25, pm       NaN
2266     G437006  03/29 07:31 am  Mar 29, 4 40, pm       NaN

[2267 rows x 4 columns]

Empty DataFrame
Columns: [incident_id, start_date, updateTime, timestamp]
Index: []

Empty DataFrame
Columns: [incident_id, start_date, updateTime, timestamp]
Index: []

Empty DataFrame
Columns: [incident_id, start_date, updateTime, timestamp]
Index: []

As you can see, only the first 4 of the 7 dataframes (which coincide with the GA providers) have null timestamps and they are all in March 2023 (this initial dataset created from 3/2023 to 1/2024).

Hence, from now on for future updates of the dataset, we will assume there are NO MORE null timestamps and the only null ones will be 3/2023

#### Transformation Before and After

Creating a transform_test function to separately try out transformations outside of the final transform function above

In [324]:
def transform_test(pipeline): # if edited, must recreate pipeline to reset transformed flag
    #### HELPER METHOD
    def _reformat_start_date(row): # taking the row and reformating its 'start_date' based on timestamp
        # Split the date string into components
        # start_date format: 03/15 05:28 pm
        month_day, time, ampm = row['start_date'].split(' ') # taking the row's start_date and parsing it
        # Split the month and day and determine year
        s_month, s_day = month_day.split('/')
        year = None
        # Determining year using timestamp as start_date does not include year
        if pd.notna(row['timestamp']): # if the timestamp value for the given row is not null (hopefully it works)
        # timestamp format: 01-18-2024 15:25:06
            timestamp_components = row['timestamp'].split(' ')
            ts_date_comp = timestamp_components[0].split('-')
            t_month, t_day = ts_date_comp[0], ts_date_comp[1]
            t_year = pd.to_numeric(ts_date_comp[2])
            if t_month == '01' and s_month == '12':
                # if the timestamp is in january but the start time is reported to be in december, make the update time have the previous year (year of Jan - 1)
                year = str(int(t_year) - 1)
            else: # use timestamp year
                year = t_year 
        else: # if no timestamp
        # for Walton, Tri-State, Oconee, and Mitchell, the na timestamps are in march 2023
            year = '2023'

        # Extract the hour and minute from the time
        hour, minute = time.split(':')

        # Convert hour to 24-hour format
        if 'am' in ampm.lower() and hour == '12': # if 12 am, then set to 00
            hour = '00' 
        if 'pm' in ampm.lower():
            hour = str(int(hour) + 12) if int(hour) < 12 else hour

        # Add leading zeros if necessary
        hour = hour.zfill(2)
        minute = minute.zfill(2)

        # Combine the components into the desired format
        reformatted_date = f'{s_month}-{s_day}-{year} {hour}:{minute}:00'

        return reformatted_date


    # Helper method
    def _reformat_update(row):
        # Format: Mar 15, 5 09, pm
        ## Splitting updateTime into components month, day, hour, min, am/pm
        month_day, time, ampm = row['updateTime'].split(',') # split into date; time; and pm/am 
        # Splitting into month, day
        u_month, u_day = month_day.split(' ')
        month_dict = { 'Jan': '01', 'Feb': '02', 'Mar': '03', 'Apr': '04', 'May': '05', 'Jun': '06', 
                    'Jul': '07', 'Aug': '08', 'Sep': '09', 'Oct': '10', 'Nov': '11', 'Dec': '12' }
        u_month = month_dict[u_month] # turning the 3 char month name to its two digit form
        
        # Determining year from timestamp as updateTime does not have year
        year = None
        if pd.notna(row['timestamp']): # if the timestamp value for the given row is not null (hopefully it works)
            # timestamp format: 01-18-2024 15:25:06
            timestamp_components = row['timestamp'].split(' ')
            ts_date_comp = timestamp_components[0].split('-')
            t_month, t_day = ts_date_comp[0], ts_date_comp[1]
            t_year = pd.to_numeric(ts_date_comp[2])

            if t_month == '01' and u_month == '12':
                # if the timestamp is in january but the start time is reported to be in december, make the update time have the previous year (year of Jan - 1)
                year = str(int(t_year) - 1)
            else: # use timestamp year
                year = t_year 
        else: # if no timestamp
        # for Walton, Tri-State, Oconee, and Mitchell, the na timestamps are in march 2023
            year = '2023'

        # Extract the hour and minute from the update time
        hour, minute = time.split() # splits based on whtie space which is space in this case and removed leading space
        # Convert hour to 24-hour format
        if 'am' in ampm.lower() and hour == '12': # if 12 am, then set to 00
            hour = '00' 
        if 'pm' in ampm.lower():
            hour = str(int(hour) + 12) if int(hour) < 12 else hour # if pm and less than 12, add 12 hours  --> othertest it is 12 pm

        # Add leading zeros if necessary
        hour = hour.zfill(2)
        minute = minute.zfill(2)

        # Combine the components into the desired format
        reformatted_date = f'{u_month}-{u_day}-{year} {hour}:{minute}:00'

        return reformatted_date

    ## HELPER METHOD ####
    

    try:
        pipeline._data['start_date'] = pipeline._data.apply(_reformat_start_date, axis=1) # reformattin the string to be datetime string form
        pipeline._data['start_date'] = pd.to_datetime(pipeline._data['start_date']) # change to timestamp data format
        pipeline._data['updateTime'] = pipeline._data.apply(_reformat_update, axis=1)
        pipeline._data['updateTime'] = pd.to_datetime(pipeline._data['updateTime'])
        pipeline._data['duration'] = pd.to_timedelta(pipeline._data['duration'])
        pipeline._data['timestamp'] = pd.to_datetime(pipeline._data['timestamp']) # THIS HAS TO BE BEFORE applying reformat functions as splitting only works on timestamp in string form, not datetime

        # Renaming column names to match with superclass standardize
        pipeline._data = pipeline._data.rename(columns={
            'incident_id':'outage_id',
            'zip_code':'zipcode'
        })


    except Exception as e:
        print(f"An error occurred during transformation: {e}")


In [325]:
print("Before transform")
print(test_p._data.info()) # test_p is from mitchell which is one of the null timestamp pipelines
display(test_p._data[['incident_id', 'start_date', 'updateTime']])
transform_test(test_p)
print("after transform")
print(test_p._data.info())
display(test_p._data[['outage_id', 'start_date', 'updateTime']])


Before transform
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123812 entries, 0 to 123811
Data columns (total 22 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   substation              109696 non-null  float64
 1   feeder                  109696 non-null  float64
 2   incident_id             123812 non-null  object 
 3   alias                   0 non-null       float64
 4   outage_comment          45473 non-null   object 
 5   estimated_restore_time  1 non-null       object 
 6   formatted_ert           1 non-null       object 
 7   start_date              123812 non-null  object 
 8   duration                123812 non-null  object 
 9   consumers_affected      123812 non-null  int64  
 10  lon                     106572 non-null  float64
 11  lat                     106572 non-null  float64
 12  opt_code                11893 non-null   object 
 13  service_index_name      123812 non-null  object 
 14  out

incident_id      start_date        updateTime
0          G437006  03/15 07:14 am  Mar 15, 3 25, pm
1          C483513  03/15 02:24 pm  Mar 15, 3 25, pm
2          G437006  03/15 07:14 am  Mar 15, 3 40, pm
3          G437006  03/15 07:14 am  Mar 15, 3 55, pm
4          G437006  03/15 07:14 am  Mar 15, 4 10, pm
...            ...             ...               ...
123807     G437006  01/17 06:32 am  Jan 18, 9 25, am
123808     D500951  01/15 07:10 am  Jan 18, 9 25, am
123809     D500998  01/17 07:20 am  Jan 18, 9 25, am
123810     C501047  01/18 09:20 am  Jan 18, 9 25, am
123811     C501018  01/17 03:26 pm  Jan 18, 9 25, am

[123812 rows x 3 columns]

after transform
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123812 entries, 0 to 123811
Data columns (total 22 columns):
 #   Column                  Non-Null Count   Dtype          
---  ------                  --------------   -----          
 0   substation              109696 non-null  float64        
 1   feeder                  109696 non-null  float64        
 2   outage_id               123812 non-null  object         
 3   alias                   0 non-null       float64        
 4   outage_comment          45473 non-null   object         
 5   estimated_restore_time  1 non-null       object         
 6   formatted_ert           1 non-null       object         
 7   start_date              123812 non-null  datetime64[ns] 
 8   duration                123812 non-null  timedelta64[ns]
 9   consumers_affected      123812 non-null  int64          
 10  lon                     106572 non-null  float64        
 11  lat                     106572 non-null  float64        
 12  

outage_id          start_date          updateTime
0        G437006 2023-03-15 07:14:00 2023-03-15 15:25:00
1        C483513 2023-03-15 14:24:00 2023-03-15 15:25:00
2        G437006 2023-03-15 07:14:00 2023-03-15 15:40:00
3        G437006 2023-03-15 07:14:00 2023-03-15 15:55:00
4        G437006 2023-03-15 07:14:00 2023-03-15 16:10:00
...          ...                 ...                 ...
123807   G437006 2024-01-17 06:32:00 2024-01-18 09:25:00
123808   D500951 2024-01-15 07:10:00 2024-01-18 09:25:00
123809   D500998 2024-01-17 07:20:00 2024-01-18 09:25:00
123810   C501047 2024-01-18 09:20:00 2024-01-18 09:25:00
123811   C501018 2024-01-17 15:26:00 2024-01-18 09:25:00

[123812 rows x 3 columns]

## Validating Data via Summary Metrics

Now that we transformed the data, we can do try to do some validtaion and calculations on the various data.

In [326]:
def _validate(group): # custom aggregating function for each groupby unique outage_id to see wehether the latest outage start date for a outage_id is before the earliest update time
    num_rows = len(group)
    num_unique_lon = len(group['lon'].unique())
    num_unique_lat = len(group['lat'].unique())
    num_unique_outages = len(group['outages'].unique())
    num_unique_customers_aff = len(group['consumers_affected'].unique())
    num_unique_customers = len(group['NumConsumers'].unique())
    num_unique_start_times = len(group['start_date'].unique())
    earliest_start_time = group['start_date'].min()
    latest_start_time = group['start_date'].max()
    earliest_update_time = group['updateTime'].min()
    latest_update_time = group['updateTime'].max()
    earliest_timestamp = group['timestamp'].min()
    latest_timestamp = group['timestamp'].max()
    duration_by_lastupt_laststrt = latest_update_time - latest_start_time
    duration_by_lastupt_firststart = latest_update_time - earliest_start_time
    duration_by_update_time = latest_update_time - earliest_update_time
    duration_by_timestamp = latest_timestamp - earliest_timestamp
    timestamp_upt_diff = latest_timestamp - latest_update_time
    is_update_dur_div_15_min = (duration_by_update_time.total_seconds()/60) % 15 == 0
    max_duration = group['duration'].max()
    update_dur_timestamp_dur_err = abs(duration_by_update_time - duration_by_timestamp) if pd.notna(duration_by_update_time) and pd.notna(duration_by_timestamp) else pd.NaT
    update_dur_max_dur_error = abs(max_duration - duration_by_update_time) if pd.notna(max_duration) and pd.notna(duration_by_update_time) else pd.NaT
    is_timestamp_dur_greater_max_dur = duration_by_timestamp > max_duration
    timestamp_dur_max_dur_err = duration_by_timestamp - max_duration if pd.notna(max_duration) and pd.notna(duration_by_timestamp) else pd.NaT
    does_start_update_intersect = latest_start_time > earliest_update_time and earliest_start_time < latest_update_time# does start and update intersect?
    is_start_after_update = earliest_start_time >= latest_update_time # is updatetime range just completely before the startdate range? 
    is_start_before_update = latest_start_time <= earliest_update_time
    is_timestamp_null = group['timestamp'].isnull().any()
    is_start_after_timestamp = earliest_start_time > earliest_timestamp and not is_timestamp_null
    is_start_before_timestamp = earliest_start_time < earliest_timestamp and not is_timestamp_null
    is_timestamp_upt_00_min = (timestamp_upt_diff.total_seconds() // 60) % 60 == 0 if pd.notna(timestamp_upt_diff) else None

    return pd.Series({
        'num_rows': num_rows,
        'num_unique_lon': num_unique_lon,
        'num_unique_lat': num_unique_lat,
        'num_unique_outages': num_unique_outages,
        'num_unique_customers_aff': num_unique_customers_aff,
        'num_unique_customers': num_unique_customers,
        'num_unique_start_times': num_unique_start_times,
        'earliest_start_time': earliest_start_time,
        'latest_start_time': latest_start_time, 
        'earliest_update_time': earliest_update_time,
        'latest_update_time': latest_update_time, 
        'earliest_timestamp': earliest_start_time,
        'latest_timestamp': latest_timestamp,
        'is_start_before_update': is_start_before_update, 
        'does_start_update_intersect': does_start_update_intersect, # is the latest start time before the earliest update time?
        'is_start_after_update': is_start_after_update, # is updatetime range just completely before the startdate range? 
        'max_duration': max_duration, # duration from the given df column
        'duration_by_update': duration_by_update_time,
        'duration_by_timestamp': duration_by_timestamp,
        'timestamp_upt_diff': timestamp_upt_diff,
        'update_dur_max_dur_error': update_dur_max_dur_error,
        'update_dur_timestamp_dur_err': update_dur_timestamp_dur_err,
        'is_timestamp_dur_greater_max_dur': is_timestamp_dur_greater_max_dur,
        'timestamp_dur_max_dur_err': timestamp_dur_max_dur_err,
        'duration_by_lastupt_laststrt': duration_by_lastupt_laststrt, # duration from calculating via update times
        'duration_by_lastupt_firststart': duration_by_lastupt_firststart, 
        'is_update_dur_div_15_min': is_update_dur_div_15_min,
        'is_timestamp_upt_00_min': is_timestamp_upt_00_min,
        'is_timestamp_null': is_timestamp_null,
        'is_start_before_timestamp': is_start_before_timestamp,
        'is_start_after_timestamp': is_start_after_timestamp

    })

In [327]:
# Testing on a dataframe
grouped = test_p._data.groupby("outage_id").apply(_validate)
display(grouped)

C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4068882227.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  grouped = test_p._data.groupby("outage_id").apply(_validate)


num_rows  num_unique_lon  num_unique_lat  num_unique_outages  \
outage_id                                                                 
C483513           1               1               1                   1   
C483521           1               1               1                   1   
C483523           6               1               1                   2   
C483525           6               1               1                   1   
C483528           1               1               1                   1   
...             ...             ...             ...                 ...   
D501048           4               1               1                   3   
G437006       27691             785             786                 274   
G484576           1               1               1                   1   
G495149           4               2               2                   4   
G495228          20               1               1                  18   

           num_unique_customers_aff  num_unique_customers  \
outage_id                                                   
C483513                           1                     1   
C483521                           1                     1   
C483523                           1                     1   
C483525                           1                     2   
C483528                           1                     1   
...                             ...                   ...   
D501048                           1                     1   
G437006                          59                   103   
G484576                           1                     1   
G495149                           2                     1   
G495228                           1                     1   

           num_unique_start_times earliest_start_time   latest_start_time  \
outage_id                                                                   
C483513                         1 2023-03-15 14:24:00 2023-03-15 14:24:00   
C483521                         1 2023-03-16 07:22:00 2023-03-16 07:22:00   
C483523                         1 2023-03-16 13:27:00 2023-03-16 13:27:00   
C483525                         1 2023-03-17 06:04:00 2023-03-17 06:04:00   
C483528                         1 2023-03-17 08:34:00 2023-03-17 08:34:00   
...                           ...                 ...                 ...   
D501048                         1 2024-01-18 09:35:00 2024-01-18 09:35:00   
G437006                       264 2023-03-15 07:14:00 2024-01-17 06:32:00   
G484576                         1 2023-04-14 10:25:00 2023-04-14 10:25:00   
G495149                         1 2023-08-30 09:40:00 2023-08-30 09:40:00   
G495228                         1 2023-08-30 11:12:00 2023-08-30 11:12:00   

          earliest_update_time  latest_update_time  earliest_timestamp  \
outage_id                                                                
C483513    2023-03-15 15:25:00 2023-03-15 15:25:00 2023-03-15 14:24:00   
C483521    2023-03-16 06:25:00 2023-03-16 06:25:00 2023-03-16 07:22:00   
C483523    2023-03-16 12:55:00 2023-03-16 14:10:00 2023-03-16 13:27:00   
C483525    2023-03-17 05:25:00 2023-03-17 06:55:00 2023-03-17 06:04:00   
C483528    2023-03-17 07:40:00 2023-03-17 07:40:00 2023-03-17 08:34:00   
...                        ...                 ...                 ...   
D501048    2024-01-18 08:40:00 2024-01-18 09:25:00 2024-01-18 09:35:00   
G437006    2023-03-15 15:25:00 2024-01-18 09:25:00 2023-03-15 07:14:00   
G484576    2023-04-14 09:55:00 2023-04-14 09:55:00 2023-04-14 10:25:00   
G495149    2023-08-30 08:55:00 2023-08-30 09:40:00 2023-08-30 09:40:00   
G495228    2023-08-30 10:40:00 2023-08-30 15:25:00 2023-08-30 11:12:00   

             latest_timestamp  is_start_before_update  \
outage_id                                               
C483513                   NaT                    True   
C483521                   NaT                   False   
C483523                   NaT                   

As we notice, while we have start times, update times, and timestamps:
- there could be multiple start times
- the range of start times could intersect or come after the range of update times which doesn't make any sense
- the timestamps are hours after the updateTimes and suspiciously in hour intervals

Also, there is a large outlier in outage_id: G437006.	

#### Let's first check to see what outages have multiple start times and see what about the instances of multiple start times.

In [328]:
df = test_p._data
# Group by outage_id and have a new temp 2nd column which counts the number of unique start_dates
grouped = df.groupby('outage_id')['start_date'].nunique()
display(grouped)

# Within grouped, finds the outage_id's that have ">1" which means more than 1 unique start_date (outage_id are indexes of grouped), and turn it into a list
filtered_ids = grouped[grouped > 1].index.tolist()

# Subset the original DataFrame based on these incident_ids
subset_df = df[df['outage_id'].isin(filtered_ids)]
print(f"{len(subset_df['outage_id'].unique())} outage_id's with more than 1 unique start date") # 25 outages with more than 1 unique start_date
# display(subset_df)
cleaned_multi_group = subset_df.groupby('outage_id')['start_date'].apply(lambda x: x.max() - x.min())
display(cleaned_multi_group)

outage_id
C483513      1
C483521      1
C483523      1
C483525      1
C483528      1
          ... 
D501048      1
G437006    264
G484576      1
G495149      1
G495228      1
Name: start_date, Length: 7344, dtype: int64

25 outage_id's with more than 1 unique start date


outage_id
C491856     0 days 00:10:00
C498646     0 days 00:51:00
C498756     0 days 02:00:00
D487369     0 days 01:24:00
D488297     1 days 19:26:00
D488463     0 days 19:21:00
D488474     1 days 13:29:00
D488539     1 days 12:03:00
D488807     0 days 10:02:00
D488816     0 days 19:35:00
D489015     1 days 11:26:00
D489020     0 days 17:52:00
D489095     0 days 17:09:00
D489097     0 days 05:25:00
D489104     0 days 22:17:00
D489142     1 days 10:18:00
D489332     1 days 18:57:00
D489462     0 days 16:43:00
D489484     0 days 18:31:00
D489493     0 days 16:28:00
D489495     0 days 18:39:00
D489497     0 days 13:06:00
D489603     0 days 00:11:00
D493173     0 days 01:00:00
G437006   307 days 23:18:00
Name: start_date, dtype: timedelta64[ns]

Let's run find all multi-start_date outageid's dataframes and keep a running list:

In [329]:
# separate code block to 1-time transform and allow the below code block be run independently
temp_pipelines1 = []
for conf in config_list:
    temp_pipelines1.append(GA11TX12(conf, base_file_path))

for pipeline in temp_pipelines1:
    pipeline.load_data()    
    transform_test(pipeline) # cre

#IMPORTANT: IF I DID temp_pipelines1 = pipelines.copy() 
#           Even though the list of pipelines are copied, since the items are references to pipelines, 
#           any changes made to pipelines in temp (like transforming) will be reflected in the original 
#           pipelines in original pipeline list

/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Walton EMC.csv


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:32: DtypeWarning: Columns (20,21) have mixed types. Specify dtype option on import or set low_memory=False.
  self._data = pd.read_csv(file_path)


/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Tri-State EMC.csv
/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Oconee EMC.csv
/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Mitchell EMC.csv


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:32: DtypeWarning: Columns (5,6,21) have mixed types. Specify dtype option on import or set low_memory=False.
  self._data = pd.read_csv(file_path)


/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/tx/layout_12/per_outage_Wise Electric Coop, Inc..csv
/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/tx/layout_12/per_outage_Houston County Electric Coop, Inc..csv
/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/tx/layout_12/per_outage_Cherokee County Electric Coop Association.csv


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:32: DtypeWarning: Columns (5,6) have mixed types. Specify dtype option on import or set low_memory=False.
  self._data = pd.read_csv(file_path)


Finding multi-start_date outage_id's

In [330]:
multi_start_list = []
df_multi = pd.DataFrame()
total_unique_outages = 0
for pipeline in temp_pipelines1:
    df = pipeline._data
    grouped = df.groupby(['outage_id'])['start_date'].nunique()
    total_unique_outages = total_unique_outages + len(grouped)
    filtered_ids = grouped[grouped > 1].index.tolist()
    multi_start_list.extend(filtered_ids)
    subset_df = df[df['outage_id'].isin(filtered_ids)]
    
    cleaned_multi_group = subset_df.groupby('outage_id')['start_date'].apply(lambda x: x.max() - x.min()) # creating a grouping for each multi-start outage where i find the diff in their latest start and their latest end
    df_multi = pd.concat([df_multi, cleaned_multi_group])

print(multi_start_list)
print(f"{len(multi_start_list)} incident id's with more than 1 unique start date out of {total_unique_outages} total unique incident_id's. {len(multi_start_list)/total_unique_outages} rate")

# rename columns and convert to datetime
df_multi = df_multi.rename(columns={'start_date': 'start_date_range'})
df_multi['start_date_range'] = pd.to_timedelta(df_multi['start_date_range'])
display(df_multi)

# Removing the G437006 outlier case

print(f"{'G437006' in multi_start_list}")

df_multi_wo_outlier = df_multi.drop('G437006').copy() # outlier of G437006 has start dates from 3/2023 to 1/2024 thus skewing the determinations
print(f"Max start_date_range: {df_multi_wo_outlier['start_date_range'].max()}")
print("Min start_date_range: ", df_multi_wo_outlier['start_date_range'].min())

# Calculating mean of date ranges
# sdr_sec = df_multi_wo_outlier['start_date_range'].dt.total_seconds() 
# avg_sec = sdr_sec.mean()
# avg_sdr = pd.Timedelta(seconds=avg_sec)
# print(f"Average start_date_range: {pd.Timedelta(seconds=(df_multi_wo_outlier['start_date_range'].dt.total_seconds()).mean())}") # change to seconds, take the mean, and convert back to timedelta
print("Average start_date_range: ", df_multi_wo_outlier['start_date_range'].mean())

['C1001443', 'C1002889', 'C1003971', 'C1009359', 'C951552', 'C952864', 'C960441', 'C961166', 'C962486', 'C979523', 'C979577', 'C980226', 'C988065', 'C991075', 'C992458', 'C996555', 'C997576', 'D1001543', 'D1003884', 'D1008494', 'D967733', 'D969201', 'D980225', 'D987758', 'D990731', 'D995253', 'D997655', 'D997960', 'G961469', 'D763983', 'D764404', 'D764738', 'D764891', 'D764938', 'D765097', 'D765100', 'D765256', 'D765957', 'D765989', 'D766098', 'D766921', 'D766995', 'D767082', 'D767114', 'D767144', 'D767276', 'C491856', 'C498646', 'C498756', 'D487369', 'D488297', 'D488463', 'D488474', 'D488539', 'D488807', 'D488816', 'D489015', 'D489020', 'D489095', 'D489097', 'D489104', 'D489142', 'D489332', 'D489462', 'D489484', 'D489493', 'D489495', 'D489497', 'D489603', 'D493173', 'G437006', 'G720140', 'G728187', 'D469964', 'D487132', 'D487135']
76 incident id's with more than 1 unique start date out of 40188 total unique incident_id's. 0.0018911117746591023 rate


start_date_range
C1001443  0 days 00:15:00
C1002889  0 days 14:50:00
C1003971  0 days 12:30:00
C1009359  0 days 16:21:00
C951552   0 days 01:00:00
...                   ...
G720140   0 days 00:29:00
G728187   0 days 00:11:00
D469964   0 days 01:27:00
D487132   0 days 01:30:00
D487135   0 days 00:44:00

[76 rows x 1 columns]

True
Max start_date_range: 6 days 00:00:00
Min start_date_range:  0 days 00:08:00
Average start_date_range:  0 days 15:13:36.800000


As we see, there are only a very few amount of incident_id's (76 out of 40188) that have more than 1 unique start_date. And among those with multiple unique start_dates, they are on average less than a day apart. 

Thus for the purposes of this data cleaning, we will use the single start_date's for our calculations, and we will remove the outages with multiple start_dates.  

## Summary of Validation Metrics across all Providers

In [331]:
# Transforming that actual pipelines
for pipeline in pipelines:
    transform_test(pipeline)

In [332]:
# Removing the multi-starttime outage-id's and testing on temp_pipelines1
for pipeline in temp_pipelines1:
    # remove outliers
    df = pipeline._data
    pipeline._data = df[~df['outage_id'].isin(multi_start_list)]

In [333]:
# Helper Function for returning a summary row for each provider as a dictionary
def summary_stats(pipeline, validated):
    subdf = {
        "Provider": pipeline.config['name'],
        "Avg num rows per outage_id": validated['num_rows'].mean(),
        "Num outages": len(validated),
        "Average unique outages per outage": validated['num_unique_outages'].mean(),
        "Average unique num cust aff": validated['num_unique_customers_aff'].mean(),
        "Average unique num cust": validated['num_unique_customers'].mean(),
        "Average unique start times": validated['num_unique_start_times'].mean(),
        "Unique Longitude Avg": validated['num_unique_lon'].mean(),
        "Unique Long Median": validated['num_unique_lon'].median(),
        "Unique Long Max": validated['num_unique_lon'].max(),
        "Unique Long Std": validated['num_unique_lon'].std(),
        "Unique Latitude Avg": validated['num_unique_lat'].mean(),
        "Unique Lat Median": validated['num_unique_lat'].median(),
        "Unique Lat Max": validated['num_unique_lat'].max(),
        "Unique Lat Std": validated['num_unique_lat'].std(),
        "Unique startDate Avg": validated['num_unique_start_times'].mean(),
        "Unique startDate Median": validated['num_unique_start_times'].median(),
        "Unique startDate Max": validated['num_unique_start_times'].max(),
        "Unique startDate Std": validated['num_unique_start_times'].std(),
        "Average update_dur_max_dur error": validated['update_dur_max_dur_error'].mean(),
        "Average update_dur_timestamp_dur error": validated['update_dur_timestamp_dur_err'].mean(),
        "Average Max Duration": validated['max_duration'].mean(),
        "Average Duration by Timestamp": validated['duration_by_timestamp'].mean(),
        "Timestamp_dur_greater_than_max_dur_rate": len(validated[validated['is_timestamp_dur_greater_max_dur'] == True]) / len(validated),
        "Average timestamp_dur_max_dur error": validated['timestamp_dur_max_dur_err'].mean(),
        "start intersects update or start > update freq": (len(validated[validated['is_start_before_update'] == False]) / len(validated)),
        "is_update_dur_div_15_min freq": len(validated[validated['is_update_dur_div_15_min']]) / len(validated),
        "is_timestamp_upt_00_min freq": len(validated[validated['is_timestamp_upt_00_min'] == True]) / len(validated),
        "is_timestamp_null": len(validated[validated['is_timestamp_null'] == True]) / len(validated),
        "is_start_before_timestamp": len(validated[validated['is_start_before_timestamp'] == True]) / len(validated),
        "is_start_after_timestamp": len(validated[validated['is_start_after_timestamp'] == True]) / len(validated)
    }
    return subdf

In [334]:
# For only single-starttime outages
summary_data = []
for pipeline in temp_pipelines1:
    validated = pipeline._data.groupby('outage_id').apply(_validate, include_groups=False).reset_index()
        # validated is the validated rows for a provider where each row is a unique outage_id
    summary_data.append(summary_stats(pipeline, validated)) # row-wise appending
display(pd.DataFrame(summary_data))

Provider  Avg num rows per outage_id  \
0                                 Walton EMC                    4.844895   
1                              Tri-State EMC                   25.023507   
2                                 Oconee EMC                    7.532451   
3                               Mitchell EMC                   12.916781   
4                   Wise Electric Coop, Inc.                    4.817185   
5         Houston County Electric Coop, Inc.                    8.471049   
6  Cherokee County Electric Coop Association                   10.778313   

   Num outages  Average unique outages per outage  \
0        14661                           3.423914   
1         1574                          12.614358   
2         7935                           6.288847   
3         7318                           4.723558   
4         2188                           2.347349   
5         1658                           5.460193   
6         4777                           8.093783   

   Average unique num cust aff  Average unique num cust  \
0                     1.003410                 1.016029   
1                     1.066074                 1.174714   
2                     1.001512                 1.012728   
3                     1.006559                 1.080896   
4                     1.002285                 1.053016   
5                     1.034982                 1.056092   
6                     1.000419                 1.116600   

   Average unique start times  Unique Longitude Avg  Unique Long Median  \
0                         1.0              1.000000                 1.0   
1                         1.0              1.003177                 1.0   
2                         1.0              1.000126                 1.0   
3                         1.0              1.000137                 1.0   
4                         1.0              1.000000                 1.0   
5                         1.0              1.029554                 1.0   
6                         1.0              1.000000                 1.0   

   Unique Long Max  Unique Long Std  Unique Latitude Avg  Unique Lat Median  \
0                1         0.000000             1.000000                1.0   
1                2         0.056290             1.003177                1.0   
2                2         0.011226             1.000126                1.0   
3                2         0.011690             1.000137                1.0   
4                1         0.000000             1.000000                1.0   
5                6         0.270844             1.028951                1.0   
6                1         0.000000             1.000000                1.0   

   Unique Lat Max  Unique Lat Std  Unique startDate Avg  \
0               1        0.000000                   1.0   
1               2        0.056290                   1.0   
2               2        0.011226                   1.0   
3               2        0.011690                   1.0   
4               1        0.000000                   1.0   
5               6        0.269793                   1.0   
6               1        0.000000                   1.0   

   Unique startDate Median  Unique startDate Max  Unique startDate Std  \
0                      1.0                     1                   0.0   
1                      1.0                     1                   0.0   
2                      1.0                     1                   0.0   
3                      1.0                     1                   0.0   
4                      1.0                     1                   0.0   
5                      1.0                     1                   0.0   
6                      1.0                     1                   0.0   

  Average update_dur_max_dur error Average update_dur_timestamp_dur error  \
0        0 days 00:32:15.375485983              0 days 00:06:51.199496362   
1        0 days 01:20:55.921219822              0 days 00:00:03.552887735  

In [335]:
# For all outages including multi-starttime outages
summary_data = []
for pipeline in pipelines:
    validated = pipeline._data.groupby('outage_id').apply(_validate, include_groups=False).reset_index()
        # validated is the validated rows for a provider where each row is a unique outage_id
    summary_data.append(summary_stats(pipeline, validated)) # row-wise appending
display(pd.DataFrame(summary_data))

Provider  Avg num rows per outage_id  \
0                                 Walton EMC                    4.860041   
1                              Tri-State EMC                   25.818353   
2                                 Oconee EMC                    7.532451   
3                               Mitchell EMC                   16.858932   
4                   Wise Electric Coop, Inc.                    4.817185   
5         Houston County Electric Coop, Inc.                    8.483735   
6  Cherokee County Electric Coop Association                   10.775523   

   Num outages  Average unique outages per outage  \
0        14690                           3.426004   
1         1591                          13.199874   
2         7935                           6.288847   
3         7344                           4.888617   
4         2188                           2.347349   
5         1660                           5.476506   
6         4780                           8.092259   

   Average unique num cust aff  Average unique num cust  \
0                     1.003540                 1.016201   
1                     1.069767                 1.179133   
2                     1.001512                 1.012728   
3                     1.014434                 1.097086   
4                     1.002285                 1.053016   
5                     1.036747                 1.056024   
6                     1.000418                 1.116527   

   Average unique start times  Unique Longitude Avg  Unique Long Median  \
0                    1.002178              1.000000                 1.0   
1                    1.011314              1.003143                 1.0   
2                    1.000000              1.000126                 1.0   
3                    1.039352              1.106890                 1.0   
4                    1.000000              1.000000                 1.0   
5                    1.001205              1.030120                 1.0   
6                    1.000628              1.000000                 1.0   

   Unique Long Max  Unique Long Std  Unique Latitude Avg  Unique Lat Median  \
0                1         0.000000             1.000000                1.0   
1                2         0.055989             1.003143                1.0   
2                2         0.011226             1.000126                1.0   
3              785         9.148502             1.107026                1.0   
4                1         0.000000             1.000000                1.0   
5                6         0.271728             1.029518                1.0   
6                1         0.000000             1.000000                1.0   

   Unique Lat Max  Unique Lat Std  Unique startDate Avg  \
0               1        0.000000              1.002178   
1               2        0.055989              1.011314   
2               2        0.011226              1.000000   
3             786        9.160171              1.039352   
4               1        0.000000              1.000000   
5               6        0.270683              1.001205   
6               1        0.000000              1.000628   

   Unique startDate Median  Unique startDate Max  Unique startDate Std  \
0                      1.0                     3              0.050816   
1                      1.0                     3              0.111582   
2                      1.0                     1              0.000000   
3                      1.0                   264              3.069569   
4                      1.0                     1              0.000000   
5                      1.0                     2              0.034700   
6                      1.0                     2              0.025047   

  Average update_dur_max_dur error Average update_dur_timestamp_dur error  \
0        0 days 00:32:28.019060585              0 days 00:06:51.464185981   
1        0 days 01:28:30.999371464              0 days 00:00:03.621309370  

Start date and End date?

From the summary table above, we notice how the 4 GA providers have a majority of their start_date ranges intersecting with their updateTime ranges (far right column) making start date a dubious metric to use for a start date. However, after eliminating the multi-start_date outages, the start_date is the best option. 

How about the duration?
- Most of the update time differences are in 15 min intervals meaning it is not really precise
- Some providers have their latest timestamp strictly some hours (ex: 6:00:00) after the latest updatetime (a non trivial %) meaning the timestamp may be dependent on the updatetime (maybe timestamp is some timezone diff away from updatetime)

Comparing max duration vs diff in update vs diff in timestamp:
- The max duration vs the difference in update yields non-trivial differences (several hours to over a day long), so that means that the difference update times could not be the entirety of the outage (ex: the first updatetime is associated with a duration meaning an outage is ongoing at this point) (ex2: outages with only 1 line meaning the diff in updatetime is 0 but either the updatetime should be the start/end of the outage).
- The max duration vs the diff in timestamp yields non-trivial differences as well.
- Curiously, diff in update and diff in timestamp show the smallest difference between them which could allude that update and timestamp are more related with each other than max duration is to either.
- With the clues that update time and timestamp may not be precise and independent, max_duration seems like a more promising candidate.

Therefore, by process of eliminating updateTime diff, timestamp diff, and starttime diff, we will use the duration column for the duration calculation

- Thus, we will use the start_date as the start and add the max_duration to the start_date to get the end date.

# Running the Pipeline

In [336]:
# Instantiate a BasePipeline object for each provider in the configuration
for provider in config['providers']:
    pipeline = GA11TX12(provider, base_file_path)
    pipeline.standardize()
    display(pipeline._data)

/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Walton EMC.csv


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:32: DtypeWarning: Columns (20,21) have mixed types. Specify dtype option on import or set low_memory=False.
  self._data = pd.read_csv(file_path)


Before transforming: len of df: 71394
After transforming: len of df: 71031


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:225: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  grouped = self._data.groupby('outage_id').apply(self._compute_metrics).reset_index().round(2)


outage_id          start_time            end_time    lat   long  \
0      C1000672 2023-11-18 18:08:00 2023-11-21 16:39:00  33.71 -84.01   
1      C1000694 2023-11-20 09:15:00 2023-11-20 11:23:00  33.78 -83.86   
2      C1000697 2023-11-20 09:41:00 2023-11-20 10:09:00  33.74 -83.70   
3      C1000701 2023-11-20 10:19:00 2023-11-20 10:24:00  33.91 -83.41   
4      C1000702 2023-11-20 11:06:00 2023-11-20 13:10:00  33.75 -83.99   
...         ...                 ...                 ...    ...    ...   
14656   G995287 2023-09-08 21:49:00 2023-09-09 00:24:00   0.00 -85.49   
14657   G996351 2023-09-12 16:20:00 2023-09-12 17:54:00   0.00 -85.49   
14658   G996585 2023-09-15 05:24:00 2023-09-15 06:09:00   0.00 -85.49   
14659   G997542 2023-09-28 21:01:00 2023-09-28 22:09:00   0.00 -85.49   
14660   G997667 2023-10-01 15:44:00 2023-10-01 17:23:00   0.00 -85.49   

       zipcode county_name  county_fips utility_provider    state  \
0        30012    Rockdale      13247.0       Walton EMC  Georgia   
1        30052      Walton      13297.0       Walton EMC  Georgia   
2        30655      Walton      13297.0       Walton EMC  Georgia   
3        30606      Clarke      13059.0       Walton EMC  Georgia   
4        30012    Rockdale      13247.0       Walton EMC  Georgia   
...        ...         ...          ...              ...      ...   
14656  unknown        None          NaN       Walton EMC     None   
14657  unknown        None          NaN       Walton EMC     None   
14658  unknown        None          NaN       Walton EMC     None   
14659  unknown        None          NaN       Walton EMC     None   
14660  unknown        None          NaN       Walton EMC     None   

        duration_diff    duration_max   duration_mean  customer_affected_mean  \
0     2 days 22:31:00 2 days 22:46:00 2 days 22:38:30                     1.0   
1     0 days 02:08:00 0 days 02:23:00 0 days 02:15:30                     1.0   
2     0 days 00:28:00 0 days 00:43:00 0 days 00:35:30                     1.0   
3     0 days 00:05:00 0 days 00:20:00 0 days 00:12:30                     1.0   
4     0 days 02:04:00 0 days 02:19:00 0 days 02:11:30                     1.0   
...               ...             ...             ...                     ...   
14656 0 days 02:35:00 0 days 02:50:00 0 days 02:42:30                    83.9   
14657 0 days 01:34:00 0 days 01:49:00 0 days 01:41:30                     3.0   
14658 0 days 00:45:00 0 days 01:00:00 0 days 00:52:30                     2.0   
14659 0 days 01:08:00 0 days 01:23:00 0 days 01:15:30                    12.0   
14660 0 days 01:39:00 0 days 01:54:00 0 days 01:46:30                     2.0   

      total_customer_outage_time  
0                2 days 22:31:00  
1                0 days 02:08:00  
2                0 days 00:28:00  
3                0 days 00:05:00  
4                0 days 02:04:00  
...                          ...  
14656            9 days 00:44:30  
14657            0 days 04:42:00  
14658            0 days 01:30:00  
14659            0 days 13:36:00  
14660            0 days 03:18:00  

[14661 rows x 15 columns]

/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Tri-State EMC.csv
Before transforming: len of df: 41077
After transforming: len of df: 39387


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:225: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  grouped = self._data.groupby('outage_id').apply(self._compute_metrics).reset_index().round(2)


outage_id          start_time            end_time    lat   long  zipcode  \
0      D760313 2023-03-16 21:34:00 2023-03-16 22:10:00  35.12 -84.33    37391   
1      D760353 2023-03-19 09:16:00 2023-03-19 17:09:00  34.93 -84.32    30559   
2      D760483 2023-03-20 01:05:00 2023-03-20 02:24:00  34.92 -84.28    30559   
3      D760513 2023-03-20 07:37:00 2023-03-20 09:24:00  34.99 -84.27    30559   
4      D760514 2023-03-20 11:50:00 2023-03-20 12:39:00  34.84 -84.30    30513   
...        ...                 ...                 ...    ...    ...      ...   
1569   D774504 2024-01-17 23:23:00 2024-01-18 01:39:00  34.90 -84.27    30559   
1570   D774507 2024-01-18 04:47:00 2024-01-18 07:23:00  35.09 -84.30  unknown   
1571   D774509 2024-01-18 04:47:00 2024-01-18 07:23:00  35.09 -84.30    28906   
1572   D774512 2024-01-18 09:07:00 2024-01-18 09:39:00  34.94 -84.28    30559   
1573   D774513 2024-01-18 08:34:00 2024-01-18 09:39:00  34.94 -84.28    30559   

     county_name  county_fips utility_provider           state  \
0           Polk      47139.0    Tri-State EMC       Tennessee   
1         Fannin      13111.0    Tri-State EMC         Georgia   
2         Fannin      13111.0    Tri-State EMC         Georgia   
3         Fannin      13111.0    Tri-State EMC         Georgia   
4         Fannin      13111.0    Tri-State EMC         Georgia   
...          ...          ...              ...             ...   
1569      Fannin      13111.0    Tri-State EMC         Georgia   
1570        None          NaN    Tri-State EMC            None   
1571    Cherokee      37039.0    Tri-State EMC  North Carolina   
1572      Fannin      13111.0    Tri-State EMC         Georgia   
1573      Fannin      13111.0    Tri-State EMC         Georgia   

       duration_diff    duration_max   duration_mean  customer_affected_mean  \
0    0 days 00:36:00 0 days 00:51:00 0 days 00:43:30                     3.0   
1    0 days 07:53:00 0 days 08:08:00 0 days 08:00:30                   780.0   
2    0 days 01:19:00 0 days 01:34:00 0 days 01:26:30                   823.0   
3    0 days 01:47:00 0 days 02:02:00 0 days 01:54:30                     1.0   
4    0 days 00:49:00 0 days 01:04:00 0 days 00:56:30                     1.0   
...              ...             ...             ...                     ...   
1569 0 days 02:16:00 0 days 02:31:00 0 days 02:23:30                     2.0   
1570 0 days 02:36:00 0 days 02:51:00 0 days 02:43:30                     2.0   
1571 0 days 02:36:00 0 days 02:51:00 0 days 02:43:30                     7.0   
1572 0 days 00:32:00 0 days 00:47:00 0 days 00:39:30                     3.0   
1573 0 days 01:05:00 0 days 01:20:00 0 days 01:12:30                     2.0   

     total_customer_outage_time  
0               0 days 01:48:00  
1             256 days 05:00:00  
2              45 days 03:37:00  
3               0 days 01:47:00  
4               0 days 00:49:00  
...                         ...  
1569            0 days 04:32:00  
1570            0 days 05:12:00  
1571            0 days 18:12:00  
1572            0 days 01:36:00  
1573            0 days 02:10:00  

[1574 rows x 15 columns]

/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Oconee EMC.csv
Before transforming: len of df: 59770
After transforming: len of df: 59770


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:225: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  grouped = self._data.groupby('outage_id').apply(self._compute_metrics).reset_index().round(2)


outage_id          start_time            end_time    lat   long  zipcode  \
0      C434216 2023-03-15 16:31:00 2023-03-15 17:24:00  32.70 -83.50    31295   
1      C434219 2023-03-16 19:42:00 2023-03-16 20:39:00  32.46 -83.18  unknown   
2      C434221 2023-03-17 15:02:00 2023-03-17 15:39:00  32.62 -83.35    31044   
3      C434222 2023-03-17 15:15:00 2023-03-17 15:39:00  32.62 -83.35    31044   
4      C434224 2023-03-17 19:15:00 2023-03-17 19:54:00  32.56 -82.95    31021   
...        ...                 ...                 ...    ...    ...      ...   
7930   D444934 2024-01-10 13:07:00 2024-01-10 14:39:00  32.61 -82.97    31021   
7931   D444941 2024-01-10 16:57:00 2024-01-10 19:23:00  32.65 -82.97  unknown   
7932   D444972 2024-01-13 11:53:00 2024-01-13 14:08:00  32.90 -83.25    31054   
7933   D444980 2024-01-16 10:10:00 2024-01-16 11:39:00  32.90 -83.25    31054   
7934   D444998 2024-01-16 19:24:00 2024-01-16 21:38:00  32.60 -83.37    31044   

     county_name  county_fips utility_provider    state   duration_diff  \
0           None          NaN       Oconee EMC     None 0 days 00:53:00   
1           None          NaN       Oconee EMC     None 0 days 00:57:00   
2         Twiggs      13289.0       Oconee EMC  Georgia 0 days 00:37:00   
3         Twiggs      13289.0       Oconee EMC  Georgia 0 days 00:24:00   
4        Laurens      13175.0       Oconee EMC  Georgia 0 days 00:39:00   
...          ...          ...              ...      ...             ...   
7930     Laurens      13175.0       Oconee EMC  Georgia 0 days 01:32:00   
7931        None          NaN       Oconee EMC     None 0 days 02:26:00   
7932   Wilkinson      13319.0       Oconee EMC  Georgia 0 days 02:15:00   
7933   Wilkinson      13319.0       Oconee EMC  Georgia 0 days 01:29:00   
7934      Twiggs      13289.0       Oconee EMC  Georgia 0 days 02:14:00   

        duration_max   duration_mean  customer_affected_mean  \
0    0 days 01:08:00 0 days 01:00:30                     1.0   
1    0 days 01:12:00 0 days 01:04:30                     1.0   
2    0 days 00:52:00 0 days 00:44:30                     1.0   
3    0 days 00:39:00 0 days 00:31:30                     1.0   
4    0 days 00:54:00 0 days 00:46:30                     1.0   
...              ...             ...                     ...   
7930 0 days 01:47:00 0 days 01:39:30                     1.0   
7931 0 days 02:41:00 0 days 02:33:30                     1.0   
7932 0 days 02:30:00 0 days 02:22:30                     4.0   
7933 0 days 01:44:00 0 days 01:36:30                     5.0   
7934 0 days 02:29:00 0 days 02:21:30                    29.0   

     total_customer_outage_time  
0               0 days 00:53:00  
1               0 days 00:57:00  
2               0 days 00:37:00  
3               0 days 00:24:00  
4               0 days 00:39:00  
...                         ...  
7930            0 days 01:32:00  
7931            0 days 02:26:00  
7932            0 days 09:00:00  
7933            0 days 07:25:00  
7934            2 days 16:46:00  

[7935 rows x 15 columns]

/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/ga/layout_11/per_outage_Mitchell EMC.csv


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:32: DtypeWarning: Columns (5,6,21) have mixed types. Specify dtype option on import or set low_memory=False.
  self._data = pd.read_csv(file_path)


Before transforming: len of df: 123812
After transforming: len of df: 94528


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:225: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  grouped = self._data.groupby('outage_id').apply(self._compute_metrics).reset_index().round(2)


outage_id          start_time            end_time    lat   long  zipcode  \
0      C483513 2023-03-15 14:24:00 2023-03-15 16:24:00  31.21 -84.37    31730   
1      C483521 2023-03-16 07:22:00 2023-03-16 07:24:00  31.18 -84.10    31779   
2      C483523 2023-03-16 13:27:00 2023-03-16 15:09:00  31.25 -84.43  unknown   
3      C483525 2023-03-17 06:04:00 2023-03-17 07:53:00  31.40 -83.72    31795   
4      C483528 2023-03-17 08:34:00 2023-03-17 08:39:00  31.24 -84.26    31730   
...        ...                 ...                 ...    ...    ...      ...   
7314   D501046 2024-01-18 09:11:00 2024-01-18 09:24:00  31.39 -84.67  unknown   
7315   D501048 2024-01-18 09:35:00 2024-01-18 10:23:00  31.39 -84.67    39813   
7316   G484576 2023-04-14 10:25:00 2023-04-14 10:55:00  31.55 -84.31    31721   
7317   G495149 2023-08-30 09:40:00 2023-08-30 10:40:00  31.55 -83.96    31791   
7318   G495228 2023-08-30 11:12:00 2023-08-30 16:24:00   0.00 -91.49  unknown   

     county_name  county_fips utility_provider    state   duration_diff  \
0       Mitchell      13205.0     Mitchell EMC  Georgia 0 days 02:00:00   
1       Mitchell      13205.0     Mitchell EMC  Georgia 0 days 00:02:00   
2           None          NaN     Mitchell EMC     None 0 days 01:42:00   
3          Worth      13321.0     Mitchell EMC  Georgia 0 days 01:49:00   
4       Mitchell      13205.0     Mitchell EMC  Georgia 0 days 00:05:00   
...          ...          ...              ...      ...             ...   
7314        None          NaN     Mitchell EMC     None 0 days 00:13:00   
7315     Calhoun      13037.0     Mitchell EMC  Georgia 0 days 00:48:00   
7316   Dougherty      13095.0     Mitchell EMC  Georgia 0 days 00:30:00   
7317       Worth      13321.0     Mitchell EMC  Georgia 0 days 01:00:00   
7318        None          NaN     Mitchell EMC     None 0 days 05:12:00   

        duration_max   duration_mean  customer_affected_mean  \
0    0 days 02:15:00 0 days 02:07:30                     1.0   
1    0 days 00:17:00 0 days 00:09:30                     1.0   
2    0 days 01:57:00 0 days 01:49:30                     1.0   
3    0 days 02:04:00 0 days 01:56:30                     1.0   
4    0 days 00:20:00 0 days 00:12:30                     1.0   
...              ...             ...                     ...   
7314 0 days 00:28:00 0 days 00:20:30                     1.0   
7315 0 days 01:03:00 0 days 00:55:30                     1.0   
7316 0 days 00:45:00 0 days 00:37:30                    45.0   
7317 0 days 01:15:00 0 days 01:07:30                     4.0   
7318 0 days 05:27:00 0 days 05:19:30                     1.0   

     total_customer_outage_time  
0               0 days 02:00:00  
1               0 days 00:02:00  
2               0 days 01:42:00  
3               0 days 01:49:00  
4               0 days 00:05:00  
...                         ...  
7314            0 days 00:13:00  
7315            0 days 00:48:00  
7316            0 days 22:30:00  
7317            0 days 04:00:00  
7318            0 days 05:12:00  

[7319 rows x 15 columns]

/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/tx/layout_12/per_outage_Wise Electric Coop, Inc..csv
Before transforming: len of df: 10540
After transforming: len of df: 10540


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:225: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  grouped = self._data.groupby('outage_id').apply(self._compute_metrics).reset_index().round(2)


outage_id          start_time            end_time    lat    long  \
0      C399381 2023-04-15 13:07:00 2023-04-15 14:16:00  33.04  -97.75   
1      C399382 2023-04-16 07:35:00 2023-04-16 09:16:00  33.06  -97.78   
2      C399386 2023-04-16 16:29:00 2023-04-16 17:46:00  33.17  -97.44   
3      C399401 2023-04-17 16:07:00 2023-04-17 18:00:00  33.28  -97.56   
4      C399402 2023-04-18 08:55:00 2023-04-18 09:31:00  33.34  -97.63   
...        ...                 ...                 ...    ...     ...   
2183   G406396 2023-12-30 07:23:00 2023-12-30 08:46:00   0.00 -103.49   
2184   G406415 2023-12-30 12:41:00 2023-12-30 13:46:00   0.00 -103.49   
2185   G406964 2024-01-12 01:31:00 2024-01-12 02:31:00   0.00 -103.49   
2186   G406967 2024-01-12 04:23:00 2024-01-12 05:47:00   0.00 -103.49   
2187   G406973 2024-01-12 04:54:00 2024-01-12 06:31:00   0.00 -103.49   

      zipcode county_name  county_fips          utility_provider  state  \
0       76083        None          NaN  Wise Electric Coop, Inc.   None   
1       76073        Wise      48497.0  Wise Electric Coop, Inc.  Texas   
2       76234        Wise      48497.0  Wise Electric Coop, Inc.  Texas   
3     unknown        None          NaN  Wise Electric Coop, Inc.   None   
4     unknown        None          NaN  Wise Electric Coop, Inc.   None   
...       ...         ...          ...                       ...    ...   
2183  unknown        None          NaN  Wise Electric Coop, Inc.   None   
2184  unknown        None          NaN  Wise Electric Coop, Inc.   None   
2185  unknown        None          NaN  Wise Electric Coop, Inc.   None   
2186  unknown        None          NaN  Wise Electric Coop, Inc.   None   
2187  unknown        None          NaN  Wise Electric Coop, Inc.   None   

       duration_diff    duration_max   duration_mean  customer_affected_mean  \
0    0 days 01:09:00 0 days 01:24:00 0 days 01:16:30                     1.0   
1    0 days 01:41:00 0 days 01:56:00 0 days 01:48:30                     1.0   
2    0 days 01:17:00 0 days 01:32:00 0 days 01:24:30                     1.0   
3    0 days 01:53:00 0 days 02:08:00 0 days 02:00:30                     1.0   
4    0 days 00:36:00 0 days 00:51:00 0 days 00:43:30                     1.0   
...              ...             ...             ...                     ...   
2183 0 days 01:23:00 0 days 01:38:00 0 days 01:30:30                    14.0   
2184 0 days 01:05:00 0 days 01:20:00 0 days 01:12:30                    12.0   
2185 0 days 01:00:00 0 days 01:15:00 0 days 01:07:30                     6.0   
2186 0 days 01:24:00 0 days 01:39:00 0 days 01:31:30                     2.0   
2187 0 days 01:37:00 0 days 01:52:00 0 days 01:44:30                     5.0   

     total_customer_outage_time  
0               0 days 01:09:00  
1               0 days 01:41:00  
2               0 days 01:17:00  
3               0 days 01:53:00  
4               0 days 00:36:00  
...                         ...  
2183            0 days 19:22:00  
2184            0 days 13:00:00  
2185            0 days 06:00:00  
2186            0 days 02:48:00  
2187            0 days 08:05:00  

[2188 rows x 15 columns]

/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/tx/layout_12/per_outage_Houston County Electric Coop, Inc..csv
Before transforming: len of df: 14083
After transforming: len of df: 14045


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:225: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  grouped = self._data.groupby('outage_id').apply(self._compute_metrics).reset_index().round(2)


outage_id          start_time            end_time    lat   long  zipcode  \
0      D712398 2023-04-14 07:21:00 2023-04-14 07:46:00  31.28 -95.39    75835   
1      D712405 2023-04-15 06:13:00 2023-04-15 07:16:00  30.93 -95.67    75852   
2      D712420 2023-04-15 09:58:00 2023-04-15 11:16:00  31.72 -95.56    75801   
3      D712431 2023-04-15 12:18:00 2023-04-15 13:01:00  31.13 -95.74    75835   
4      D712434 2023-04-15 17:44:00 2023-04-15 19:16:00  31.53 -95.89  unknown   
...        ...                 ...                 ...    ...    ...      ...   
1653   G740838 2024-01-12 06:55:00 2024-01-12 07:32:00  31.22 -95.39  unknown   
1654   G741154 2024-01-12 07:48:00 2024-01-12 08:58:00  31.26 -94.89  unknown   
1655   G741203 2024-01-12 08:08:00 2024-01-12 09:55:00  31.21 -95.04  unknown   
1656   G741890 2024-01-15 05:40:00 2024-01-15 09:46:00  31.02 -95.43  unknown   
1657   G742007 2024-01-16 11:22:00 2024-01-16 12:16:00  31.44 -95.59  unknown   

     county_name  county_fips                    utility_provider  state  \
0        Houston      48225.0  Houston County Electric Coop, Inc.  Texas   
1        Madison      48313.0  Houston County Electric Coop, Inc.  Texas   
2       Anderson      48001.0  Houston County Electric Coop, Inc.  Texas   
3        Houston      48225.0  Houston County Electric Coop, Inc.  Texas   
4           None          NaN  Houston County Electric Coop, Inc.   None   
...          ...          ...                                 ...    ...   
1653        None          NaN  Houston County Electric Coop, Inc.   None   
1654        None          NaN  Houston County Electric Coop, Inc.   None   
1655        None          NaN  Houston County Electric Coop, Inc.   None   
1656        None          NaN  Houston County Electric Coop, Inc.   None   
1657        None          NaN  Houston County Electric Coop, Inc.   None   

       duration_diff    duration_max   duration_mean  customer_affected_mean  \
0    0 days 00:25:00 0 days 00:40:00 0 days 00:32:30                    2.00   
1    0 days 01:03:00 0 days 01:18:00 0 days 01:10:30                   31.00   
2    0 days 01:18:00 0 days 01:33:00 0 days 01:25:30                   44.00   
3    0 days 00:43:00 0 days 00:58:00 0 days 00:50:30                    8.00   
4    0 days 01:32:00 0 days 01:47:00 0 days 01:39:30                    7.00   
...              ...             ...             ...                     ...   
1653 0 days 00:37:00 0 days 00:52:00 0 days 00:44:30                   48.00   
1654 0 days 01:10:00 0 days 01:25:00 0 days 01:17:30                   77.00   
1655 0 days 01:47:00 0 days 02:02:00 0 days 01:54:30                    8.00   
1656 0 days 04:06:00 0 days 04:21:00 0 days 04:13:30                   68.14   
1657 0 days 00:54:00 0 days 01:09:00 0 days 01:01:30                   38.75   

     total_customer_outage_time  
0               0 days 00:50:00  
1               1 days 08:33:00  
2               2 days 09:12:00  
3               0 days 05:44:00  
4               0 days 10:44:00  
...                         ...  
1653            1 days 05:36:00  
1654            3 days 17:50:00  
1655            0 days 14:16:00  
1656 11 days 15:23:08.571428571  
1657            1 days 10:52:30  

[1658 rows x 15 columns]

/Users/uirja/OneDrive/Personal Files/CS Projects/outage-data-scraper/raw_data/tx/layout_12/per_outage_Cherokee County Electric Coop Association.csv


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:32: DtypeWarning: Columns (5,6) have mixed types. Specify dtype option on import or set low_memory=False.
  self._data = pd.read_csv(file_path)


Before transforming: len of df: 51507
After transforming: len of df: 51488


C:\Users\uirja\AppData\Local\Temp\ipykernel_10248\4133122531.py:225: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  grouped = self._data.groupby('outage_id').apply(self._compute_metrics).reset_index().round(2)


outage_id          start_time            end_time    lat   long  zipcode  \
0      C480035 2023-09-24 22:28:00 2023-09-25 17:31:00  32.16 -95.45    75762   
1      C480066 2023-09-24 22:43:00 2023-09-25 17:31:00  32.15 -95.45    75762   
2      C480094 2023-09-24 22:45:00 2023-09-25 12:16:00  32.15 -95.39    75757   
3      C480111 2023-09-24 22:46:00 2023-09-25 17:31:00  32.16 -95.45    75762   
4      C480121 2023-09-24 22:47:00 2023-09-26 05:46:00   0.00 -97.49  unknown   
...        ...                 ...                 ...    ...    ...      ...   
4772   G483387 2023-09-24 22:55:00 2023-09-26 06:01:00   0.00 -97.49  unknown   
4773   G483389 2023-09-24 23:05:00 2023-09-26 06:01:00   0.00 -97.49  unknown   
4774   G483390 2023-09-24 22:47:00 2023-09-26 06:01:00   0.00 -97.49  unknown   
4775   G483391 2023-09-24 23:04:00 2023-09-26 06:01:00   0.00 -97.49  unknown   
4776   G483392 2023-09-24 23:14:00 2023-09-26 06:01:00   0.00 -97.49  unknown   

     county_name  county_fips                           utility_provider  \
0          Smith      48423.0  Cherokee County Electric Coop Association   
1          Smith      48423.0  Cherokee County Electric Coop Association   
2       Cherokee      48073.0  Cherokee County Electric Coop Association   
3          Smith      48423.0  Cherokee County Electric Coop Association   
4           None          NaN  Cherokee County Electric Coop Association   
...          ...          ...                                        ...   
4772        None          NaN  Cherokee County Electric Coop Association   
4773        None          NaN  Cherokee County Electric Coop Association   
4774        None          NaN  Cherokee County Electric Coop Association   
4775        None          NaN  Cherokee County Electric Coop Association   
4776        None          NaN  Cherokee County Electric Coop Association   

      state   duration_diff    duration_max   duration_mean  \
0     Texas 0 days 19:03:00 0 days 19:18:00 0 days 19:10:30   
1     Texas 0 days 18:48:00 0 days 19:03:00 0 days 18:55:30   
2     Texas 0 days 13:31:00 0 days 13:46:00 0 days 13:38:30   
3     Texas 0 days 18:45:00 0 days 19:00:00 0 days 18:52:30   
4      None 1 days 06:59:00 1 days 07:14:00 1 days 07:06:30   
...     ...             ...             ...             ...   
4772   None 1 days 07:06:00 1 days 07:21:00 1 days 07:13:30   
4773   None 1 days 06:56:00 1 days 07:11:00 1 days 07:03:30   
4774   None 1 days 07:14:00 1 days 07:29:00 1 days 07:21:30   
4775   None 1 days 06:57:00 1 days 07:12:00 1 days 07:04:30   
4776   None 1 days 06:47:00 1 days 07:02:00 1 days 06:54:30   

      customer_affected_mean total_customer_outage_time  
0                        1.0            0 days 19:03:00  
1                        1.0            0 days 18:48:00  
2                        1.0            0 days 13:31:00  
3                        1.0            0 days 18:45:00  
4                        1.0            1 days 06:59:00  
...                      ...                        ...  
4772                    14.0           18 days 03:24:00  
4773                    18.0           23 days 04:48:00  
4774                     9.0           11 days 17:06:00  
4775                    12.0           15 days 11:24:00  
4776                     5.0            6 days 09:55:00  

[4777 rows x 15 columns]